### PR TITLE
feat(mimo): add Whisper audio encoder to LLaVA MIMO training pipeline

### PR DESCRIPTION
**Pull request by kamran-nvidia**
# What does this PR do ?

Adds end-to-end audio modality support to the MIMO LLaVA example, alongside the existing CLIP vision branch. Specifically:

# Changelog

- **Whisper encoder**: new [examples/models/megatron_mimo/whisper/](examples/models/megatron_mimo/whisper/) subpackage with a Megatron-style Whisper encoder ([whisper_model.py](examples/models/megatron_mimo/whisper/whisper_model.py), [whisper_layer_specs.py](examples/models/megatron_mimo/whisper/whisper_layer_specs.py)).
- **HF → Megatron conversion + verification**: [convert_hf_whisper_to_megatron.py](examples/models/megatron_mimo/whisper/convert_hf_whisper_to_megatron.py) emits per-TP-rank checkpoints; [verify_whisper_conversion.py](examples/models/megatron_mimo/whisper/verify_whisper_conversion.py) checks numerical parity against HuggingFace Whisper.
- **Audio-aware MIMO training script**: [megatron_mimo_training_llava_audio.py](examples/models/megatron_mimo/megatron_mimo_training_llava_audio.py) wires the Whisper encoder into the MIMO LLaVA stack (vision + audio + Vicuna-7B LLM), with matching launch scripts ([run_hetero_llava_audio.sh](examples/models/megatron_mimo/run_hetero_llava_audio.sh), [run_hetero_llava_audio_parallelism_tests.sh](examples/models/megatron_mimo/run_hetero_llava_audio_parallelism_tests.sh)).
- **Dataset preparation**: [synthesize_llava_pretrain_audio.py](examples/models/megatron_mimo/synthesize_llava_pretrain_audio.py) synthesizes TTS audio (NeMo FastPitch + HiFiGAN) for LLaVA-Pretrain and emits an augmented JSON; [prepare_llava_pretrain_audio.sh](examples/models/megatron_mimo/prepare_llava_pretrain_audio.sh) shards/merges the run.
- **Dataset hook**: small change in [src/megatron/bridge/data/megatron_mimo/dataset.py](src/megatron/bridge/data/megatron_mimo/dataset.py) to surface the new `audio` field.

Also from this PR (https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3531): 

- Adds an opt-in `--deterministic` flag to the MIMO LLaVA and LLaVA-audio
  training scripts (`megatron_mimo_training_llava.py`,
  `megatron_mimo_training_llava_audio.py`). When set, vision, language, audio,
  and projection configs run in FP32 with unfused attention, loss fusion
  disabled, and full activation recompute so runs are bitwise reproducible.
- Threads the `deterministic` argument through the relevant config and model
  spec helpers so each sub-module is configured consistently.
- Updates the launcher shell scripts under
  `examples/models/megatron_mimo/` (`run_hetero_llava.sh`,
  `run_hetero_llava_audio.sh`, and the three `*_parallelism_tests*.sh`
  variants) to forward `--deterministic` and export the env vars required for
  deterministic execution (e.g. cuBLAS / NCCL settings).
- No behavior change when the flag is omitted — default training paths are
  untouched.

# GitHub Actions CI

See the [CI section](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md#-running-github-ci)in the Contributing doc for how to trigger the CI. A Nvidia developer will need to approve and trigger the CI for external contributors.

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?

If you haven't finished some of the above items you can still open "Draft" PR.

# Additional Information

- Related to # (issue)

### CODE REVIEW

**Comment by kamran-nvidia** [comment]
/ok to test [9c558ff](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3520/commits/9c558ffcdb62097d674166ae162a0a29ce13f401)

**Comment by kamran-nvidia** [comment]
/ok to test [05f22f5](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3520/commits/05f22f54be3eaa35486eccff761e6e5738a1e865)

**Comment by kamran-nvidia** [comment]
/ok to test [3a72a2b](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3520/commits/3a72a2b9230dc57c33c783345e49c268c5e47f89)

**Comment by kamran-nvidia** [comment]
/ok to test [ab9a70e](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3520/commits/ab9a70eaaf395e9c1e81e3c261b8905a82723b8d)

**Comment by kamran-nvidia** [comment]
/ok to test [67a106e](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3520/commits/67a106e17e9db74fb409ab1d19576f4ebec2a3c5)

**Comment by kamran-nvidia** [comment]
/ok to test [cc9e68e](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3520/commits/cc9e68ee24f5cc6a417444d418603c58c0bcd837)

**Comment by kamran-nvidia** [comment]
/ok to test [4c8f254](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3520/commits/4c8f254e913da564ec929f094495718d16948cf7)

**Comment by kamran-nvidia** [comment]
/ok to test [42fb555](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3520/commits/42fb555aaa972995a12f7a54d992b5e6c0b174e5)

**Comment by kamran-nvidia** [comment]
/ok to test [00db946](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3520/commits/00db9460383296dbdb012d46f30b0aad7af5d46f)

**Comment by kamran-nvidia** [comment]
/ok to test [f9e3fe4](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3520/commits/f9e3fe43cbd01ffba70cf2cea445058acb08c434)

**Comment by kamran-nvidia** [comment]
/ok to test [368324f](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3520/commits/368324f0a9fc8ef889791808f0fba9ac411015b1)

**Comment by kamran-nvidia** [comment]
/ok to test [35140ae](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3520/commits/35140aeb76b75ce8f63e5b768930feb7a7c1cf16)

File: tests/functional_tests/test_groups/training/test_pretrain_megatron_mimo.py [hunk 1]
```diff
@@ -260,13 +288,13 @@ def _build_config(
     train_cfg.num_microbatches = 1
```
Comment (cuichenx) [inline tests/functional_tests/test_groups/training/test_pretrain_megatron_mimo.py]: is this still used? looks like a leftover phantom assignment

File: examples/models/megatron_mimo/whisper/whisper_model.py
Comment (cuichenx) [inline examples/models/megatron_mimo/whisper/whisper_model.py:1]: should this modeling code live in the examples folder?

File: tests/functional_tests/test_groups/training/test_pretrain_megatron_mimo.py (hunk 1 above)
Comment (kamran-nvidia) [reply tests/functional_tests/test_groups/training/test_pretrain_megatron_mimo.py]: Removed. thx

File: examples/models/megatron_mimo/whisper/whisper_model.py
Comment (kamran-nvidia) [reply examples/models/megatron_mimo/whisper/whisper_model.py:1]: @cuichenx Good question. I guess I can move this to the models folder but we are not qualifying the Whisper model as a stand-alone model in the MBridge with this PR. Wdyt?

Comment (liding-nv) [reply examples/models/megatron_mimo/whisper/whisper_model.py:1]: @cuichenx what's the move for where modeling code lives?
I saw it in several places.. CLIP is in MCore, QWen3VL is in MBridge. Any thoughts on where we should drop this audio encoder? As Kamran mentioned, it won't be a standalone model...

**Comment by kamran-nvidia** [comment]
/ok to test [1fc43e1](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3520/commits/1fc43e13306f7c72db6113f8ea99def9ce5b1ba5)

**Comment by kamran-nvidia** [comment]
/ok to test [43c65a7](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3520/commits/43c65a7047180ed58ba7457006bece009ea60617)

**Review by liding-nv — APPROVED** [review]
